File: erlang-projects/factorial_recursive_tco.erl
-module(factorial_recursive_tco).
-export([factorial/1]).

factorial(N) when N >= 0 ->
    factorial(N, 1).

factorial(0, Acc) ->
    Acc;
factorial(N, Acc) when N > 0 ->
    factorial(N - 1, N * Acc).

File: haskell-projects/factorial-recursive-tco.hs
factorialTCO :: Integer -> Integer
factorialTCO n = factorialHelper n 1
  where
    factorialHelper 0 acc = acc
    factorialHelper m acc = factorialHelper (m - 1) (m * acc)


main :: IO ()
main = do
  let result = factorialTCO 5
  putStrLn $ "Factorial of 5 is: " ++ show result
